File: backend/resources/import.go
package resources

import (
	"time"

	"github.com/stashsphere/backend/models"
)

type ImportStatus struct {
	ID             string     `json:"id"`
	Status         string     `json:"status"`
	CreatedAt      time.Time  `json:"createdAt"`
	CompletedAt    *time.Time `json:"completedAt"`
	Error          *string    `json:"error"`
	ThingsImported *int       `json:"thingsImported"`
	ListsImported  *int       `json:"listsImported"`
	ImagesImported *int       `json:"imagesImported"`
}

func ImportStatusFromModel(i *models.Import) ImportStatus {
	return ImportStatus{
		ID:             i.ID,
		Status:         string(i.Status),
		CreatedAt:      i.CreatedAt,
		CompletedAt:    i.CompletedAt.Ptr(),
		Error:          i.ErrorMSG.Ptr(),
		ThingsImported: i.ThingsImported.Ptr(),
		ListsImported:  i.ListsImported.Ptr(),
		ImagesImported: i.ImagesImported.Ptr(),
	}
}

File: frontend/src/api/import.ts
import { Axios } from 'axios';

export type ImportStatus = {
  id: string;
  status: string;
  createdAt: string;
  completedAt: string | null;
  error: string | null;
  thingsImported: number | null;
  listsImported: number | null;
  imagesImported: number | null;
};

export const queueImport = async (axios: Axios, file: File): Promise<ImportStatus> => {
  const formData = new FormData();
  formData.append('file', file);
  const response = await axios.post('/import', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data as ImportStatus;
};

export const getImportStatus = async (axios: Axios): Promise<ImportStatus | null> => {
  const response = await axios.get('/import', {
    validateStatus: (status) => status === 200 || status === 404,
  });
  if (response.status === 404) return null;
  return response.data as ImportStatus;
};
